HeliPort/Appearance/WiFiConfigWindow.swift

```diff
@@ -0,0 +1,591 @@
+        passwdSecureBox.stringValue = ""
+        usernameBox.stringValue = ""
+    }
+
+    func show() {
+        makeKeyAndOrderFront(self)
+    }
+
+    override func close() {
+        // checks if window was shown as a sheet
+        if let sheet = sheetParent {
+            sheet.endSheet(self, returnCode: .cancel)
+        } else {
+            super.close()
+        }
+    }
+}
+
+// MARK: NSTextFieldDelegate
+
+extension WiFiConfigWindow: NSTextFieldDelegate {
+
+    func controlTextDidChange(_ obj: Notification) {
+        // if clear password box is visible, copy password to secure box
+        if !passwdInputBox.isHidden {
+            passwdSecureBox.stringValue = passwdInputBox.stringValue
+        }
+
+        // if clear password box is not visible, copy password from secure box to password box
+        if passwdInputBox.isHidden {
+            passwdInputBox.stringValue = passwdSecureBox.stringValue
+        }
+
+        // trim secure box to 64 characters
+        if passwdSecureBox.stringValue.count > 64 {
+            passwdSecureBox.stringValue = String(passwdSecureBox.stringValue[..<passwdSecureBox.stringValue.index(
+                passwdSecureBox.stringValue.startIndex,
+                offsetBy: 64
+            )])
+        }
+
+        // trim password box to 64 characters
+        if passwdInputBox.stringValue.count > 64 {
+            passwdInputBox.stringValue = String(passwdInputBox.stringValue[..<passwdInputBox.stringValue.index(
+                passwdInputBox.stringValue.startIndex,
+                offsetBy: 64
+            )])
+        }
+
+        controlJoinButton()
+    }
+
+    private func controlJoinButton() {
+        // SSID needs to be filled in and shorter than 32 characters
+        guard !networkBox.stringValue.isEmpty, networkBox.stringValue.count <= 32 else {
+            rightButton.isEnabled = false
+            return
+        }
+
+        // no password used, both password inputs are hidden
+        if passwdInputBox.isHidden, passwdSecureBox.isHidden {
+            rightButton.isEnabled = true
+            return
+        }
+
+        // password is too short, less than 8 characters
+        guard (!passwdInputBox.isHidden || !passwdSecureBox.isHidden),
+            passwdSecureBox.stringValue.count >= 8,
+            passwdInputBox.stringValue.count >= 8  else {
+            rightButton.isEnabled = false
+            return
+        }
+
+        // user name input shown but not filled in
+        if !usernameBox.isHidden, usernameBox.stringValue.isEmpty {
+            rightButton.isEnabled = false
+            return
+        }
+
+        // everything is OK
+        rightButton.isEnabled = true
+    }
+}
+
+// MARK: Action Items
+
+extension WiFiConfigWindow {
+
+    @objc private func security(_ sender: Any?) {
+        switch securityPop.title {
+        case .none:
+            gridView.row(at: .usernameRow).isHidden = true
+            gridView.row(at: .passwordRow).isHidden = true
```

Comment:
should clear up `passwdInputBox.stringValue` and `passwdSecureBox.stringValue.count` here.

Reply:
After the switch case, it calls `resetInputBoxes()` and `controlJoinButton()` on lines 447 and 448, which clears out password and username and resets the `joinButton` state. However I can also just move those two methods in each switch case.

Reply:
Oops, forget about it. I just missed that line.